File: 2-addTwoNumbers.php
<?php
class ListNode {
     public $val = 0;
     public $next = null;
     function __construct($val = 0, $next = null) {
         $this->val = $val;
         $this->next = $next;
     }
}

function addTwoNumbers ($l1, $l2)
{
    // 使用$l1作为结果链表，$res记录指向链表的开始地方，防止开始指针丢失
    $res = $l1;
    $addFlag = 0; // 进位标志
    $sum = 0;
    while ($l1 != null && $l2 != null) {
        $sum = $l1->val + $l2->val + $addFlag;
        $addFlag = floor($sum / 10);
        $l1->val = $sum % 10;

        // 防止指针走空
        if (empty($l1->next) || empty($l2->next)) break;
        $l1 = $l1->next;
        $l2 = $l2->next;
    }

    if ($l1->next == null && $l2->next != null) {
        $l2 = $l2->next; // $l2在上面已经计算过，所以指向next先再处理
        while ($l2) {
            if ($addFlag == 0) {
                // 如果没有进位了，则直接把l2链表后的元素补到l1即可
                $l1->next = $l2;
                break;
            } else {
                $sum = $l2->val + $addFlag;
                $addFlag = floor($sum / 10 ) ;
                $val = $sum % 10;
                var_dump($val);

                // 构造新的链表节点挂到 l1 上
                $l1->next = new ListNode($val);
                $l1 = $l1->next;
                $l2 = $l2->next;

            }
        }
    }
    if ($l1->next != null && $l2->next == null) {
        $l1 = $l1->next;
        while($l1 && $addFlag == 1) {
            $sum = $l1->val + $addFlag;
            $addFlag = floor($sum / 10);
            $val = $sum % 10;
            $l1->val = $val;
            // 防止指向为空
            if (empty($l1->next)) break;
            $l1 = $l1->next;
        }
    }
    if ($addFlag == 1) {
        $l1->next = new ListNode(1);
    }

    return $res;
}

File: go-code/2-addTwoNumbers.go
package main

/**
 * Definition for singly-linked list.
 * type ListNode struct {
 *     Val int
 *     Next *ListNode
 * }
 */

func addTwoNumbers(l1 *ListNode, l2 *ListNode) *ListNode {
	res := l1
	addFlag := 0
	sum := 0
	for l1 != nil && l2 != nil {
		sum = l1.Val + l2.Val + addFlag
		if sum > 9 {
			addFlag = 1
		} else {
			addFlag = 0
		}
		l1.Val = sum % 10

		if l1.Next == nil || l2.Next == nil {
			break
		}
		l1 = l1.Next
		l2 = l2.Next
	}
	if l1.Next == nil && l2.Next != nil {
		l2 = l2.Next
		for l2 != nil {
			if addFlag == 0 {
				l1.Next = l2
				break
			} else {
				sum = l2.Val + addFlag
				if sum > 9 {
					addFlag = 1
				} else {
					addFlag = 0
				}
				val := sum % 10

				l1.Next = &ListNode{
					Val:  val,
					Next: nil,
				}
				l1 = l1.Next
				l2 = l2.Next
			}
		}
	}

	if l1.Next != nil && l2.Next == nil {
		l1 = l1.Next
		for l1 != nil && addFlag == 1 {
			sum = l1.Val + addFlag
			if sum > 9 {
				addFlag = 1
			} else {
				addFlag = 0
			}
			val := sum % 10

			l1.Val = val
			if l1.Next == nil {
				break
			}
			l1 = l1.Next
		}
	}

	if addFlag == 1 {
		l1.Next = &ListNode{
			Val:  addFlag,
			Next: nil,
		}
	}

	return res
}
